fdbserver/workloads/AuthzSecurity.actor.cpp

```diff
@@ -0,0 +1,409 @@
+		actorCount = getOption(options, "actorsPerClient"_sr, transactionsPerSecond);
+		tenant = getOption(options, "tenantA"_sr, "authzSecurityTestTenant"_sr);
+		anotherTenant = getOption(options, "tenantB"_sr, "authzSecurityTestTenant"_sr);
+		tLogConfigKey = getOption(options, "tLogConfigKey"_sr, "TLogInterface"_sr);
+		ASSERT(g_network->isSimulated());
+		// make it comfortably longer than the timeout of the workload
+		signedToken = g_simulator->makeToken(
+		    tenant, uint64_t(std::lround(getCheckTimeout())) + uint64_t(std::lround(testDuration)) + 100);
+		signedTokenAnotherTenant = g_simulator->makeToken(
+		    anotherTenant, uint64_t(std::lround(getCheckTimeout())) + uint64_t(std::lround(testDuration)) + 100);
+		testFunctions.push_back(
+		    [this](Database cx) { return testCrossTenantGetDisallowed(PositiveTestcase::True, cx, this); });
+		testFunctions.push_back(
+		    [this](Database cx) { return testCrossTenantGetDisallowed(PositiveTestcase::False, cx, this); });
+		testFunctions.push_back(
+		    [this](Database cx) { return testCrossTenantCommitDisallowed(PositiveTestcase::True, cx, this); });
+		testFunctions.push_back(
+		    [this](Database cx) { return testCrossTenantCommitDisallowed(PositiveTestcase::False, cx, this); });
+		testFunctions.push_back(
+		    [this](Database cx) { return testPublicNonTenantRequestsAllowedWithoutTokens(cx, this); });
+		testFunctions.push_back([this](Database cx) { return testTLogReadDisallowed(cx, this); });
+	}
+
+	Future<Void> setup(Database const& cx) override { return Void(); }
+	Future<Void> start(Database const& cx) override {
+		for (int c = 0; c < actorCount; c++)
+			clients.push_back(timeout(runTestClient(cx, this), testDuration, Void()));
+		return waitForAll(clients);
+	}
+	Future<bool> check(Database const& cx) override {
+		int errors = 0;
+		for (int c = 0; c < clients.size(); c++)
+			errors += clients[c].isError();
+		if (errors)
+			TraceEvent(SevError, "TestFailure").detail("Reason", "There were client errors.");
+		clients.clear();
+		return errors == 0 && crossTenantGetPositive.getValue() > 0 && crossTenantGetNegative.getValue() > 0 &&
+		       crossTenantCommitPositive.getValue() > 0 && crossTenantCommitNegative.getValue() > 0 &&
+		       publicNonTenantRequestPositive.getValue() > 0;
+	}
+	void getMetrics(std::vector<PerfMetric>& m) override {
+		m.push_back(crossTenantGetPositive.getMetric());
+		m.push_back(crossTenantGetNegative.getMetric());
+		m.push_back(crossTenantCommitPositive.getMetric());
+		m.push_back(crossTenantCommitNegative.getMetric());
+		m.push_back(publicNonTenantRequestPositive.getMetric());
+		m.push_back(tLogReadNegative.getMetric());
+	}
+
+	void setAuthToken(Transaction& tr, Standalone<StringRef> token) {
+		tr.setOption(FDBTransactionOptions::AUTHORIZATION_TOKEN, token);
+	}
+
+	ACTOR Future<Void> setKeyValue(Transaction* tr, StringRef key, StringRef value) {
```

Comment:
Can you change the name of this to make it clear that this is also going to commit. The whole concept of this actor is a bit weird as you set a value to a transaction that we might've read from or written to, then attempt to commit, and if the commit fails it will retry only with the last set... This means the following is undefined behavior:

```
state Transaction tr(cx);
tr.set("val1"_sr, "foo"_sr);
wait(setKeyValue(&tr, "val2"_sr, "foo"_sr));
```

My preference would be not to have this actor really...

Reply:
Right. After all the iterations it has gone through, it makes little sense now that we pass a transaction.
I should make it an actor that use its own transaction instead.